File: rl-competition/system/testing/consoleParamTesterJava/src/consoleTrainer.java
/* console Trainer for RL Competition
* Copyright (C) 2007, Brian Tanner brian@tannerpages.com (http://brian.tannerpages.com/)
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA. */
import rlglue.RLGlue;

public class consoleTrainer {

//A better example of an actual experiment is to call the runMountainCarExperiment() method
	public static void main(String[] args) throws InterruptedException {
		runMountainCarExperiment();
	}
	
	//A sample program to run through the different mountain car training values
	public static void runMountainCarExperiment(){
		for(int whichParamSet=0;whichParamSet<35;whichParamSet++){
			System.out.println("Param Set: "+whichParamSet);
			consoleTrainerHelper.loadMountainCar(whichParamSet);
			runExperiment();
		}
	}
	
	private static void runExperiment(){
		RLGlue.RL_init();
		
		int episodesToRun=500;
		int trialsToRun=10;
		
		int totalSteps=0;
		int last50EpisodeTotal=0;
		
		long startTime=System.currentTimeMillis();
		for(int i=0;i<trialsToRun;i++){
			int[] stepCounts=runEpisodes(episodesToRun,100000);
			totalSteps+=stepCounts[0];
			last50EpisodeTotal+=stepCounts[1];
		}
		long endTime=System.currentTimeMillis();
		
		long totalTime=endTime-startTime;
		
		long seconds=totalTime/1000;
		
		long stepsPerSecond=totalSteps/seconds;
		double averageSteps=(double)totalSteps/(double)(episodesToRun*trialsToRun);
		double averageLast50Steps=(double)last50EpisodeTotal/(double)(50*trialsToRun);
		System.out.printf("Average steps per episode averaged over %d trials = %.2f.\tLast 50 episodes is %.2f.\tTook %d seconds or %d steps per second \n",trialsToRun,averageSteps,averageLast50Steps,seconds,stepsPerSecond);

		RLGlue.RL_cleanup();
		
	}

	private static int[] runEpisodes(int episodeCount, int maxEpisodeLength){
		int[] stepCounts=new int[2];
		int maxSteps=0;
		int minSteps=100000000;
		
		for(int i=0;i<episodeCount;i++){
			RLGlue.RL_episode(maxEpisodeLength);
			
			int thisSteps=RLGlue.RL_num_steps();
			if(thisSteps>maxSteps)maxSteps=thisSteps;
			if(thisSteps<minSteps)minSteps=thisSteps;
			stepCounts[0]+=thisSteps;
			if(episodeCount-i<=50)
				stepCounts[1]+=thisSteps;
		}
		return stepCounts;
	}
	

	

}

File: branches/rl-competition-RC1/trainers/consoleTrainerCPP/src/consoleTrainer.cpp
/* console Trainer for RL Competition
* Copyright (C) 2007, Brian Tanner brian@tannerpages.com (http://brian.tannerpages.com/)
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA. */
#include <iostream>
#include <RL_glue.h>

#include "consoleTrainerHelper.h"

int main(int argc, char *argv[])
{
	int whichTrainingMDP = 0;
	// Uncomment ONE of the following lines to choose your experiment
	//loadTetris(whichTrainingMDP); //put the desired parameter set in where the 0 is
	loadMountainCar(whichTrainingMDP); //put the desired parameter set in where the 0 is
	//loadHelicopter();

	RL_init();

	int episodeCount=10; //number of episodes to run
	int maxEpisodeLength=20000; //set a maxEpisodeLength to ensure termination (RL_glue defaults to 100,000)
		
	int totalSteps=0;//counter for the total number of steps taken to finish all episodes
	//run the episodes with RL_episode(maxEpisodeLength)
	for(int i=0;i<episodeCount;i++){
		RL_episode(maxEpisodeLength);
		std::cout<<"Episode: "<<i<<" steps: "<<RL_num_steps()<<std::endl;//output to see how many steps the episode took
		totalSteps+=RL_num_steps();
	}
	std::cout<<"Total Steps : "<<totalSteps<<std::endl;		
	
	//clean up the environment and end the program
	RL_cleanup();

    return 0;
}
